Implement hypervolume history plot for matplotlib backend

y0z opened the pull request:
## Motivation
This PR implements a hypervolume history plot function to visualize the achieved hypervolume at each trial to evaluate the performance of a sampler in multiobjective optimization for the matplotlib backend.


### Demo
```python
import optuna

import matplotlib.pyplot as plt


def objective(trial):
    x = trial.suggest_float("x", 0, 5)
    y = trial.suggest_float("y", 0, 3)

    v0 = 4 * x ** 2 + 4 * y ** 2
    v1 = (x - 5) ** 2 + (y - 5) ** 2
    return v0, -v1


study = optuna.create_study(directions=["minimize", "maximize"])
study.optimize(objective, n_trials=50)

reference_point=[140., -50.]
optuna.visualization.matplotlib.plot_pareto_front(study)
plt.scatter(x=reference_point[0], y=reference_point[1], color='k', marker='*')
plt.tight_layout()
plt.show()

optuna.visualization.matplotlib.plot_hypervolume_history(study, reference_point)
plt.tight_layout()
plt.show()
```

<img src="https://github.com/optuna/optuna/assets/30489874/81e9a008-877f-4396-8a1a-38f83ac6bd25" width="45%">
<img src="https://github.com/optuna/optuna/assets/30489874/f17d5961-216a-44b3-b9ce-715c105445a7" width="45%">

## Description of the changes
- Add _hypervolume_history.py

not522 reviewed (commented):
Thank you for proposing the new feature! It is commonly used in multi-objective optimization and is great to introduce in the visualization module.
I left some comments for implementation. Please check them.
The calculation time for hv history may be very long, but making it a future task is no problem. (I think that we can reduce it by an incremental algorithm.)

y0z commented on optuna/visualization/matplotlib/_hypervolume_history.py:
```diff
@@ -0,0 +1,180 @@
+    study: Study,
+    reference_point: Sequence[float],
+    *,
+    constraints_func: Callable[[FrozenTrial], Sequence[float]] | None = None,
```
Thank you for your helpful feedback.
I've changed the interface for constraint handling by using `trial.system_attrs[optuna.samplers._base._CONSTRAINTS_KEY]`.

HideakiImamura commented on optuna/visualization/matplotlib/_hypervolume_history.py:
```diff
@@ -0,0 +1,160 @@
+           ``study.n_objectives`` must be 2 or more.
+
+        reference_point:
+            A reference point to use for hypervolume computation.
```
It would be great to document the length of the reference point must be same as the number of objectives.

y0z replied:
done :)

HideakiImamura commented on optuna/visualization/matplotlib/_hypervolume_history.py:
```diff
@@ -0,0 +1,160 @@
+
+    _imports.check()
+
+    assert study._is_multi_objective(), (
```
It's better to raise ValueError for this, I think.

y0z replied:
done

HideakiImamura commented on optuna/visualization/matplotlib/_hypervolume_history.py:
```diff
@@ -0,0 +1,160 @@
+        "Study must be multi-objective. For single-objective optimization, "
+        "please use plot_optimization_history instead."
+    )
+
```
It would be nice to validate the input `reference_point` (the length must be same as #objectives).

y0z replied:
done

HideakiImamura reviewed (commented):
Thanks for the PR. Basically, LGTM. I have some comments, so please take a look.

By the way,
- Could you add plot_hypervolume_history to the https://github.com/optuna/optuna/blob/master/docs/source/reference/visualization/matplotlib.rst?
- Could you try to add unit-tests? It would be worthwhile just to add a simple test for normal cases.

y0z commented:
Thank you for your comments.

> Could you add plot_hypervolume_history to the https://github.com/optuna/optuna/blob/master/docs/source/reference/visualization/matplotlib.rst?

OK. I've updated the .rst file.
https://github.com/optuna/optuna/pull/4748/files#diff-c5ae50c9546c48c4dd5d0d7567c0052ff4ea684ddde61853316e165483a347c3

> Could you try to add unit-tests? It would be worthwhile just to add a simple test for normal cases.

Sure. I've added brief tests that validate:
- the consistency with respect to minimization and maximization,
- infeasible trials never increase the hypervolume.

not522 reviewed (approved):
Thank you for your update! LGTM!

HideakiImamura reviewed (approved):
Thanks for your quick update! LGTM!
